src/pool/mempool.rs

```diff
@@ -19,97 +30,148 @@ pub type TransactionOrdering = CoinbaseTipOrdering<EthPooledTransaction>;
+                        .await
+                        .map_err(|e| eyre::eyre!("Error converting block ID: {:?}", e))?;
+
+                    // Query the initial account_nonce for the account from the provider
+                    accounts.insert(
+                        felt_address,
+                        eth_client
+                            .eth_provider()
+                            .starknet_provider()
+                            .get_nonce(starknet_block_id, felt_address)
+                            .await
+                            .unwrap_or_default(),
+                    );
                 }
             }
         }
 
-        Self { accounts }
+        if accounts.is_empty() {
+            return Err(eyre::eyre!("No accounts found in file"));
+        }
+
+        Ok(Self { accounts: Arc::new(Mutex::new(accounts)), eth_client })
     }
 
-    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: &'static EthClient<SP>)
-    where
-        SP: starknet::providers::Provider + Send + Sync + Clone + 'static,
-    {
+    /// Starts the account manager task that periodically checks account balances and processes transactions.
+    #[allow(clippy::significant_drop_tightening)]
+    pub fn start(&'static self, rt_handle: &Handle) {
         let accounts = self.accounts.clone();
 
         rt_handle.spawn(async move {
             loop {
-                for address in &accounts {
-                    Self::process_transaction(address, eth_client);
+                // Get account addresses first without acquiring the lock
+                let account_addresses: Vec<Felt> = {
+                    let accounts = accounts.lock().await;
+                    accounts.keys().copied().collect()
+                };
+
+                // Iterate over account addresses and check balances
+                let mut errors = vec![];
+                for account_address in account_addresses {
+                    match self.get_balance(&account_address).await {
+                        Ok(balance) => {
+                            if balance > U256::from(u128::pow(10, 18)) {
+                                // Acquire lock only when necessary to modify account state
+                                let mut accounts = accounts.lock().await;
+                                if let Some(account_nonce) = accounts.get_mut(&account_address) {
+                                    self.process_transaction(&account_address, account_nonce);
+                                }
+                            }
+                        }
+                        Err(e) => {
+                            tracing::error!("Error getting balance for account_address {:?}: {:?}", account_address, e);
+                            errors.push(e);
+                        }
+                    }
+                }
+
+                if !errors.is_empty() {
+                    tracing::error!("Error checking balances: {:?}", errors);
                 }
```

Comment:
aren't you tracing the errors twice?

Reply:
I still don't really understand why we can just do something more functional, like
```
let balance = self.get_balance(&account_address).await.inspect_err(|err| tracing::error!(...)).unwrap_or_default();
if balance > ... {
   ...
}
```
that way, if we have an error, we trace it, and we unwrap the balance to 0, meaning we won't use the account

Reply:
Ah yes just pushed the modification. As you mentioned the `?` in your previous comment I was thinking you thought about propagating the error which is difficult to do in the async loop. But via this yeah, the balance check will fail as unwrap to default and so we will not use the account